kapitan/inputs/kadet.py

```diff
@@ -0,0 +1,245 @@
+
+        # These will be updated per target
+        # XXX At the moment we have no other way of setting externals for modules...
+        global search_paths
+        search_paths = self.search_paths
+        global inventory
+        inventory = lambda: Dict(inventory_func(self.search_paths, target_name)) # noqa E731
+        global inventory_global
+        inventory_global = lambda: Dict(inventory_func(self.search_paths, None)) # noqa E731
+
+        kadet_module, spec = module_from_path(file_path)
+        sys.modules[spec.name] = kadet_module
+        spec.loader.exec_module(kadet_module)
+        logger.debug('Kadet.compile_file: spec.name: %s', spec.name)
+
+        output_obj = kadet_module.main().to_dict()
+        if prune:
+            output_obj = prune_empty(output_obj)
+
+        for item_key, item_value in output_obj.items():
+            # write each item to disk
+            if output == 'json':
+                file_path = os.path.join(compile_path, '%s.%s' % (item_key, output))
+                with CompiledFile(file_path, self.ref_controller, mode="w", reveal=reveal, target_name=target_name,
+                                  indent=indent) as fp:
+                    fp.write_json(item_value)
+            elif output == 'yaml':
+                file_path = os.path.join(compile_path, '%s.%s' % (item_key, "yml"))
+                with CompiledFile(file_path, self.ref_controller, mode="w", reveal=reveal, target_name=target_name,
+                                  indent=indent) as fp:
+                    fp.write_yaml(item_value)
+            else:
+                raise ValueError('output is neither "json" or "yaml"')
+            logger.debug("Pruned output for: %s", file_path)
+
+    def default_output_type(self):
+        return "yaml"
+
+
+class BaseObj(object):
+    def __init__(self, **kwargs):
+        """
+        returns a BaseObj
+        kwargs will be save into self.kwargs
+        values in self.root are returned as dict via self.to_dict()
+        """
+        self.root = Dict()
+        self.kwargs = Dict(kwargs)
+        self.new()
+        self.body()
+
+    @classmethod
+    def from_json(cls, file_path):
+        """
+        returns a BaseObj initialised with json content
+        from file_path
+        """
+        with open(file_path) as fp:
+            json_obj = json.load(fp)
+            return cls.from_dict(json_obj)
+
+    @classmethod
+    def from_yaml(cls, file_path):
+        """
+        returns a BaseObj initialised with yaml content
+        from file_path
+        """
+        with open(file_path) as fp:
+            yaml_obj = yaml.safe_load(fp)
+            return cls.from_dict(yaml_obj)
+
+    @classmethod
+    def from_dict(cls, dict_value):
+        """
+        returns a BaseObj initialise with dict_value
+        """
+        bobj = cls()
+        bobj.root = Dict(dict_value)
+        return bobj
+
+    def update_root_yaml(self, file_path):
```

Comment:
Another nitpick but could this function and update_root_json be consolidated to one? The only difference is the yaml.safe_load or json.load line. Maybe with an extra argument, or infer the type from the filename?

Reply:
good point. Fixed